pkg/provisioner/ironic/ironic.go

```diff
@@ -1346,3 +1356,51 @@ func (p *ironicProvisioner) PowerOff() (result provisioner.Result, err error) {
 
 	return result, nil
 }
+
+// SoftPowerOff tries to perform soft shutdown first
+// If soft shutdown is unspported, it performs hard shutdown
+func (p *ironicProvisioner) SoftPowerOff() (result provisioner.Result, err error) {
+	p.log.Info("ensuring host is soft powered off")
+
+	ironicNode, err := p.findExistingHost()
+	if err != nil {
+		return result, errors.Wrap(err, "failed to find existing host")
+	}
+
+	if ironicNode.PowerState != powerOff {
+		// When SoftPowerOff() is called again, if there is error on changing
+		// power status from 'power off' by 'soft power off', the error is stored
+		// in the LastError field of the the node
+		if ironicNode.LastError != "" && strings.HasPrefix(ironicNode.LastError, "Failed to change power state to 'power off' by 'soft power off'") {
+			err = errors.New(ironicNode.LastError)
+			result.RequeueAfter = powerRequeueDelay
+			return result, errors.Wrap(err, "failed to soft power off host")
+		}
+
+		if ironicNode.TargetPowerState == powerOff || ironicNode.TargetPowerState == softPowerOff {
+			p.log.Info("waiting for power status to change")
+			result.RequeueAfter = powerRequeueDelay
+			return result, nil
+		}
+		result, err = p.changePower(ironicNode, nodes.SoftPowerOff)
+		if err != nil {
+			// Soft power off is not supported by vendor driver, uses PowerOff()
+			if strings.HasPrefix(err.Error(), "driver does not support target power state") {
+				return p.PowerOff()
+			} else {
+				result.RequeueAfter = powerRequeueDelay
+				return result, errors.Wrap(err, "failed to soft power off host")
```

Comment:
Errors are retried immediately. If this case is something that might mean no retry is reasonable or that a retry should be delayed, the error message should be stored in the result so it can be attached to the host later.